File: cpp/src/0009_PalindromeNumber.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <cassert>

using namespace std;

class PalindromeNumberSolution
{
public:
    bool isPalindrome(int x)
    {
        if (x < 0 || x % 10 == 0 && x != 0)
            return false;

        uint64_t reversed = 0;
        int input, digit;
        input = x;

        do
        {
            digit = x % 10;
            x = x / 10;
            reversed = reversed * 10 + digit;
        } while (x != 0);

        return input == reversed;
    }

    void runTests()
    {
        vector<tuple<int, bool>> tests{
            {12344321, true},
            {-1221, false},
            {0, true}};

        for (int i = 0; i < tests.size(); ++i)
            assert(isPalindrome(get<0>(tests[i]) == get<1>(tests[i])));
    }
};

File: csharp/src/0009_PalindromeNumber.cs
namespace csharp.src;

public record struct PalindromeNumberSolution
{
    public static bool IsPalindrome(int x)
    {
        if (x < 0 || x % 10 == 0 && x != 0)
        {
            return false;
        }

        var input = x;
        var reversed = 0;

        do
        {
            var digit = x % 10;
            x /= 10;
            reversed = reversed * 10 + digit;
        } while (x != 0);

        return input == reversed;
    }
}

public class PalindromeNumberTests
{
    [Theory]
    [InlineData(12344321, true)]
    [InlineData(111, true)]
    [InlineData(-1221, false)]
    [InlineData(0, true)]
    public void Tests(int testNumer, bool expected)
    {
        Assert.Equal(expected, PalindromeNumberSolution.IsPalindrome(testNumer));
    }
}
